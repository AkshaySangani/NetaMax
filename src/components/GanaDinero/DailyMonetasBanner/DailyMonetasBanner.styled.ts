import styled from "styled-components";

import { Box, Button, Flex, Text } from "@chakra-ui/react";
import { responsive } from "styled/responsive";

export const MainContainer = styled(Box)`
  background: radial-gradient(#418edd, #2253c9);
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25), inset -22px -10px 71px rgba(116, 0, 0, 0.25);
  z-index: 2;
  width: ${(props) => (props.$hasScrolled ? "100%" : "90%")};
  margin: ${(props) => (props.$hasScrolled ? "0px" : "30px 5% 0px 5%")};
  border-radius: ${(props) => (props.$hasScrolled ? "0px 0px" : "12px")};
  height: ${(props) => (props.$hasScrolled ? "40px" : "70px")};
  ${responsive.mdPlus} {
    border-radius: 0px;
    width: 100%;
    margin: 0px;
    height: 50px;
  }
`;

export const BannerContainer = styled(Flex)`
  position: sticky;
  flex-direction: row;
  align-items: center;
  height: 100%;
  justify-content: center;
  margin: ${(props) => (props.$hasScrolled ? "0px 10px" : "0px")};
  ${responsive.mdPlus} {
    margin: ${(props) => (props.$hasScrolled ? "0px 30px" : "0px")};
  }
`;

export const CoinImagesHolder = styled(Flex)`
  flex-direction: column;
  height: 100%;
  justify-content: space-around;
`;

export const BannerButton = styled(Button)`
  border-radius: 10px;
  background-color: #fff;
  color: #3870ff;
  margin-left: 2.5px;
  margin-right: 5px;
  font-size: 14px;
  ${responsive.mdPlus} {
    font-size: 18px;
  }
`;

export const BannerText = styled(Text)`
  color: #fff;
  font-weight: 600;
  margin-left: 5px;
  margin-right: 2.5px;
  text-align: center;
`;
