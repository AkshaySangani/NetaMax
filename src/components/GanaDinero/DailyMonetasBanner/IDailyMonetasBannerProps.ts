export interface IDailyMonetasBannerProps {
  /**
   * the category id.
   * @type {string}
   */
  isLoggedIn: boolean;

  /**
   * the discovery challenge description
   * @type {string |undefined}
   */
  challengeDescription: string | undefined;
}
