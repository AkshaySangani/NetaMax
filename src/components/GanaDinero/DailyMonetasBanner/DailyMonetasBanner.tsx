import React, { ReactElement, useEffect, useState } from "react";

import { useMediaQuery, Image } from "@chakra-ui/react";

import { useRouter } from "next/router";

import throttle from "lodash.throttle";
import {
  BannerButton,
  BannerContainer,
  BannerText,
  CoinImagesHolder,
  MainContainer,
} from "./DailyMonetasBanner.styled";
import { screenSizes } from "styled/screen";
import { IDailyMonetasBannerProps } from "./IDailyMonetasBannerProps";

/**
 * Daily monetas banner which changes depending on if the user is logged in or not
 * @param {IDailyMonetasBanner} props  the props received which tell us if a user is logged in or not
 * @returns {ReactElement} the react element to render.
 */
export const DailyMonetasBanner = (props: IDailyMonetasBannerProps): ReactElement => {
  const router = useRouter();
  const [hasScrolled, setHasScrolled] = useState(false);
  const [scrollTop, setScrollTop] = useState(0);
  const [isLargerThanSm] = useMediaQuery(`(min-width: ${screenSizes.mdMin}px)`);

  const { isLoggedIn, challengeDescription } = props;
  /**
   * Action on item click.
   * @param {string} item the item clicked.
   * @returns {void}
   */
  const redirectTo = (item: string): void => {
    router.push(item);
  };

  const handleScroll = throttle(() => {
    const offset = document.getElementById("header")?.offsetHeight || 0;
    const { scrollTop } = document.documentElement;
    setScrollTop(scrollTop);

    const scrolled = scrollTop >= offset;
    if (hasScrolled !== scrolled) {
      setHasScrolled(scrolled);
    }
  }, 250);

  useEffect(() => {
    document.addEventListener("scroll", handleScroll);
    return () => {
      document.removeEventListener("scroll", handleScroll);
    };
  }, [hasScrolled]);

  return (
    <MainContainer
      id="gana-monetas-banner"
      $hasScrolled={hasScrolled}
      $scrollTop={scrollTop}
      sx={{
        position: "sticky",
        top: `${isLargerThanSm ? "117px" : "135px"}`,
      }}
    >
      <BannerContainer $hasScrolled={hasScrolled}>
        {!hasScrolled || isLargerThanSm ? (
          <CoinImagesHolder alignItems="flex-end" width="40px">
            <Image src="/assets/images/banner-monetas/moneda-izquierda-arriba.webp" height="20px" />
            <Image src="/assets/images/banner-monetas/moneda-izquierda-abajo.webp" height="17px" />
          </CoinImagesHolder>
        ) : null}

        <BannerText
          data-testid="daily-monetas-banner-text"
          fontSize={isLargerThanSm ? "18px" : "14px"}
        >
          {isLoggedIn && challengeDescription != undefined
            ? `${challengeDescription}! ðŸ˜œ`
            : "Â¿Quieres ganar dinero para tus compras? ðŸ˜œ"}
        </BannerText>

        <BannerButton
          onClick={() => redirectTo("./gana-dinero")}
          size={isLargerThanSm ? "sm" : hasScrolled ? "xs" : "md"}
        >
          Checa aquÃ­
        </BannerButton>
        {!hasScrolled || isLargerThanSm ? (
          <CoinImagesHolder alignItems="flex-start" width="35px">
            <Image src="/assets/images/banner-monetas/moneda-derecha-arriba.webp" height="17px" />
            <Image src="/assets/images/banner-monetas/moneda-derecha-abajo.webp" height="15px" />
          </CoinImagesHolder>
        ) : null}
      </BannerContainer>
    </MainContainer>
  );
};
