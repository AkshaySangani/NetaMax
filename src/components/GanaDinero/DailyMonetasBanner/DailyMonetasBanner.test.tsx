import { render, cleanup, screen } from "@testing-library/react";
import { DailyMonetasBanner } from "./DailyMonetasBanner";
import { IDailyMonetasBannerProps } from "./IDailyMonetasBannerProps";

describe("Render DailyMonetasBanner component", () => {
  const notLoggedInProps: IDailyMonetasBannerProps = {
    isLoggedIn: false,
    challengeDescription: undefined,
  };

  const loggedInProps: IDailyMonetasBannerProps = {
    isLoggedIn: true,
    challengeDescription: "Ven diario y junta 3 monetas ",
  };

  test("DailyMonetasBanner should render properly when a user is not logged in", () => {
    const { container } = render(<DailyMonetasBanner {...notLoggedInProps} />);
    expect(container).toMatchSnapshot();
    expect(screen.getByTestId("daily-monetas-banner-text").textContent).toMatch(
      "Â¿Quieres ganar dinero para tus compras? ğŸ˜œ"
    );
  });

  test("DailyMonetasBanner should render properly when a user is logged in", () => {
    const { container } = render(<DailyMonetasBanner {...loggedInProps} />);
    expect(container).toMatchSnapshot();
    expect(screen.getByTestId("daily-monetas-banner-text").textContent).toMatch(
      `${loggedInProps.challengeDescription}`
    );
  });

  afterEach(() => {
    cleanup();
  });
});
