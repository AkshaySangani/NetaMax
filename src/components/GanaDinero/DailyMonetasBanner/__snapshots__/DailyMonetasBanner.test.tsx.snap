// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render DailyMonetasBanner component DailyMonetasBanner should render properly when a user is logged in 1`] = `
<div>
  <div
    class="sc-bdvvtL jdjvsT css-1wld30"
    id="gana-monetas-banner"
  >
    <div
      class="sc-gsDKAQ cXlaqt css-k008qs"
    >
      <div
        class="sc-dkPtRN eCljL css-1mydh6f"
      >
        <img
          class="chakra-image css-1pg8bi"
          src="/assets/images/banner-monetas/moneda-izquierda-arriba.webp"
        />
        <img
          class="chakra-image css-10xjgmf"
          src="/assets/images/banner-monetas/moneda-izquierda-abajo.webp"
        />
      </div>
      <p
        class="chakra-text sc-eCImPb dZBwrz css-1kuy7z7"
        data-testid="daily-monetas-banner-text"
      >
        Ven diario y junta 3 monetas ! ðŸ˜œ
      </p>
      <button
        class="chakra-button sc-hKwDye jCqsQU css-5c2176"
        type="button"
      >
        Checa aquÃ­
      </button>
      <div
        class="sc-dkPtRN eCljL css-12ak7jr"
      >
        <img
          class="chakra-image css-10xjgmf"
          src="/assets/images/banner-monetas/moneda-derecha-arriba.webp"
        />
        <img
          class="chakra-image css-uanaa8"
          src="/assets/images/banner-monetas/moneda-derecha-abajo.webp"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`Render DailyMonetasBanner component DailyMonetasBanner should render properly when a user is not logged in 1`] = `
<div>
  <div
    class="sc-bdvvtL jdjvsT css-1wld30"
    id="gana-monetas-banner"
  >
    <div
      class="sc-gsDKAQ cXlaqt css-k008qs"
    >
      <div
        class="sc-dkPtRN eCljL css-1mydh6f"
      >
        <img
          class="chakra-image css-1pg8bi"
          src="/assets/images/banner-monetas/moneda-izquierda-arriba.webp"
        />
        <img
          class="chakra-image css-10xjgmf"
          src="/assets/images/banner-monetas/moneda-izquierda-abajo.webp"
        />
      </div>
      <p
        class="chakra-text sc-eCImPb dZBwrz css-1kuy7z7"
        data-testid="daily-monetas-banner-text"
      >
        Â¿Quieres ganar dinero para tus compras? ðŸ˜œ
      </p>
      <button
        class="chakra-button sc-hKwDye jCqsQU css-5c2176"
        type="button"
      >
        Checa aquÃ­
      </button>
      <div
        class="sc-dkPtRN eCljL css-12ak7jr"
      >
        <img
          class="chakra-image css-10xjgmf"
          src="/assets/images/banner-monetas/moneda-derecha-arriba.webp"
        />
        <img
          class="chakra-image css-uanaa8"
          src="/assets/images/banner-monetas/moneda-derecha-abajo.webp"
        />
      </div>
    </div>
  </div>
</div>
`;
